src/FsCheck/Gen.fs

```diff
@@ -180,13 +180,13 @@ module Gen =
     ///given probabilities. The sum of the probabilities must be larger than zero.
     //[category: Creating generators from generators]
     [<CompiledName("Frequency")>]
-    let frequency xs = 
-        let xs = Seq.cache xs
-        let tot = Seq.sumBy fst xs
-        let rec pick n ys = 
-            let (k,x),xs = Seq.head ys,Seq.skip 1 ys
-            if n<=k then x else pick (n-k) xs
-        in gen.Bind(choose (1,tot), fun n -> pick n xs) 
+    let frequency xs =
+        let xs = Seq.toArray xs
+        let tot = Array.sumBy fst xs
```

Comment:
As sumBy is evaluating the full sequence any it is often more efficient to store the result as an array. Then pick iterates over this array using array indices.